Compose formatting buttons popover

amanagr opened the pull request:
![Peek 2021-06-20 03-58](https://user-images.githubusercontent.com/25124304/122673248-2a2d5380-d1ed-11eb-99e3-9660b5e9dfae.gif)


discussion: https://chat.zulip.org/#narrow/stream/101-design/topic/compose.20box.20formatting.20button

timabbott commented on static/js/popover_menus.js:
```diff
@@ -87,4 +89,74 @@ export function initialize() {
+                    const formatted_text = format_text(range, format_type);
+                    // Replace formatted text in textbox.
+                    $textbox.val(pre_text + formatted_text + post_text);
+                    $textbox.get(0).setSelectionRange(range.start, range.end);
```
This replacement logic should probably be a function in `compose_ui.js`; that's our common module for manipulation of the compose text areas.  (And having that logic all in one place makes it a lot easier to try to audit for bugs with breaking `Ctrl+Z` or with how whitespace is managed in the compose box, since the code for such manipulations would all be in one place).

Also, I suspect that `compose.handle_keydown` is essentially a duplicate of this code; would it make sense to extract that into functions in `compose_ui.js` and then have this and that both call the same functions?

amanagr replied:
Oh! I wasn't aware of `compose.handle_keydown`. There is definitely code duplication here.

timabbott commented on static/js/util.js:
```diff
@@ -354,3 +354,126 @@ export function filter_by_word_prefix_match(items, search_term, item_to_text) {
+// then get the position of text in that div.
+// We cannot get the position of caret in textarea
+// directly because it is a replaced element
+// (https://developer.mozilla.org/en-US/docs/Web/CSS/Replaced_element).
```
I'm not sure we can use this code, since I don't see clear documentation of an open source license for it.  I'm not sure if there's a way to ask the author, but we might want to look for an OSS library that implements this function anyway.

amanagr replied:
Replaced with https://github.com/deshiknaves/caret-pos which seems like a decent library.

timabbott commented:
This looks nice!  Posted a batch of comments on implementation details.

amanagr commented:
Updated with requested changes.

alya commented:
A couple of comments from playing around with this:

1. My strong expectation from other products is that if you highlight text that already has some formatting applied (bold or italic), clicking "B" or "I" will *undo* the existing formatting. I think we need to have our buttons work the same way.

It's less clear to me what double-applying a link should do. I notice that it doesn't do anything smart in GitHub.

2. It's easy to accidentally select a blank line at the top or bottom of the selected area, making the formatting not work. E.g.:

`**
foo**`

Could we help out the user by automatically leaving out the blank line from the formatted selection?

alya commented:
I also think placing the buttons a bit to the right rather than the left might look better when the selected text is at the very left of the line? We might need to make the placement variable to fine-tune this.

Looking at Dropbox Paper, they put the formatting buttons in all sorts of places -- from a couple of minutes of playing around with it, it's not obvious what the algorithm is or why they do it the way they do.

amanagr commented:
Nice suggestions. Working on them.

amanagr commented:
I did the first part. I am still thinking the best way to do part 2.

amanagr commented:
I addressed your comments. You can adjust the position of popover from [here](https://github.com/zulip/zulip/pull/18922/commits/32b720a5e3bee15f4e6d58092300083fdf60ec8f#diff-d5c37a0dfce5be1bc2a4e0746bacfd75967a0edf7708b27b37dbe2bddb64140fR111). We can start adding more formatting buttons, once this is merged.

alya commented:
Awesome, looks good to me -- thanks!

timabbott commented on static/js/compose_ui.js:
```diff
@@ -203,67 +196,51 @@ export function handle_keydown(event, textarea) {
+
+            range.start = range.end + 3;
+            range.end = range.start + 3;
+            field.setSelectionRange(range.start, range.end);
```
How does this logic work?  I don't super understand either the magic 3s or why we're swapping end/start.

amanagr replied:
Added some comments.

timabbott replied:
OK; I did some variable name renames so that it now reads:
```
+        case "link":
+            // Ctrl + L: Insert a link to selected text
+            wrapSelection(field, "[", "](url)");
+
+            // Change selected text to `url` part of the syntax.
+            // If <text> marks the selected region, we're mapping:
+            // <text> => [text](<url>).
+            const new_start = range.end + "[](".length;
+            const new_end = new_start + "url".length;
+            field.setSelectionRange(new_start, new_end);
+            break;
+    }
```

which clearly indicates where those 3s come from, and also avoid unnecessarily mutating the `range` variable.

andersk replied:
@timabbott Your edition adds an ESLint [`no-case-declaration`](https://eslint.org/docs/rules/no-case-declarations) error. You need braces around a case that includes variable declarations.

timabbott commented on static/js/compose_ui.js:
```diff
@@ -203,67 +196,51 @@ export function handle_keydown(event, textarea) {
-            }
-        }
-
+    if (type && isCmdOrCtrl) {
```
Nice refactoring here to remove a lot of the duplication.

timabbott commented on package.json:
```diff
@@ -70,6 +70,7 @@
     "stacktrace-gps": "^3.0.4",
     "style-loader": "^2.0.0",
     "terser-webpack-plugin": "^4.1.0",
+    "text-field-edit": "^3.1.0",
```
The `text-field-edit` library doesn't seem  super popular (only 176 downloads/week) but seems maintained has a clean API that handles the undo issues here, so seems like a good fit.

timabbott commented on frontend_tests/node_tests/compose_ui.js:
```diff
@@ -470,130 +474,59 @@ run_test("markdown_shortcuts", ({override}) => {
-        queryCommandEnabled = false;
-        compose_ui.handle_keydown(event, $("#compose-textarea"));
+        assert.equal(format_text_type, "link");
+        format_text_type = undefined;
```
Thanks for doing this refactoring on the tests; it deletes a lot of message testing code.

timabbott commented on package.json:
```diff
@@ -23,6 +23,7 @@
     "babel-plugin-formatjs": "^10.2.6",
     "blueimp-md5": "^2.10.0",
     "cache-loader": "^4.0.0",
+    "caret-pos": "^2.0.0",
```
We already have the similar `jquery-caret-plugin` library.  Neither seems particularly popular; I wonder if we need a library for this or should just be using browser APIs for this (since this seems like a very common thing to want, and thus something there's likely either a much more popular library or just a clean browser API for).  @andersk FYI for the question.

amanagr replied:
I tried out `jquery-caret-plugin`, it doesn't return the `position` of cursor that I need, which is returned by `caret-pos`. Also, `caret-pos` was last updated last year while `jquery-caret-plugin` hasn't been updated in 5years.

timabbott replied:
Yeah, I think removing a library in the other direction might make more sense, especially since we do want to remove jquery dependencies where it's convenient long term.

I'm OK with merging this for now and addressing the problem of too many libraries as a follow-up.

timabbott commented on static/js/click_handlers.js:
```diff
@@ -653,7 +653,17 @@ export function initialize() {
+            });
+        } else {
+            popovers.hide_all();
+        }
```
Could we just pass `not_hide_tippy_instances` unconditionally in this code path?  

I also wonder if we want a different name/semantics for controlling that hiding logic, e.g. a more generic `except_popover: {popover_name}`?  Not sure.

amanagr replied:
Yeah, done the expect_popover thing.

timabbott replied:
I think this might not work; in that `popovers.hide_all` seems to not pass through the parameter properly.  Probably needs a preparatory refactoring to correct that.

timabbott commented on static/templates/formatting_buttons_popover.hbs:
```diff
@@ -0,0 +1,11 @@
+        <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABoAAAAaCAYAAACpSkzOAAAABmJLR0QA/wD/AP+gvaeTAAAA60lEQVRIieWVuw5BQRRFl0dDJyEKjUoiEgWNTpRq0Sip+Az/5dXwARoVFYkEiYLOoyKYI+a4k1uwk2kmM3vNrLnJhX9NDbgoxuK1IGgJKisPNlSuvycCFIAucMK8wQZoATkg9i3kNRMB1HFV/piVAMq4hmQFyNJmo+3HcEtFmOv5BRooOz4mAKwx1aVdg/ICZG67WaNO0tb3C+T8fULADlNdyjWoKEBmmgJbdZ61+QaySRjY86ztDCRdg0qY7zPVltioqwtzxh/US+JAGzhi3mgLNICEF0A ...
+    </span>
+    <span class="formatting_button">
+        <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABoAAAAaCAYAAACpSkzOAAAABmJLR0QA/wD/AP+gvaeTAAAB30lEQVRIia3Wv2sUQRjG8c/dRUQbRawVogYUjHp97LUQFMTO/AliGXKEJApaKFgKEkgsxMYfWNgIFoJFRKNRLGyvE5MyYHGixdyR2WF3cuvlhYHded95v/vOPvPuNoxuLVzFZUygh3U8xrtdyE8/8Vf8rRgr2DsqpI3NDGQwlqExAuQNDkVzH/FSqGAaRyLf+f+FpJXcEd7VwA6jG/kf7QZkviJ2MYpZbdaAHMAVYXu+9OcWMqA/0XWrImYoO5XxjWHNdkVPxoZI2MBFXBLk/Fs4JyuZNbd ...
```
What's the thinking behind having tiny PNGs here rather than using icons?  (Does Font Awesome not have the icons we need?  We also could think about moving these to our `static/assets/icons` icon font).

amanagr replied:
Font awesome doesn't have good icons for these in v3.  Using icons leads to link icons appearing weird 
![image](https://user-images.githubusercontent.com/25124304/132333534-31305600-88e7-4d43-bbd6-e20493429826.png)

I have tried several.

timabbott replied:
OK -- we should probably just move these icons to live in a system intended to store images/icons, not inline in the HTML, then.

timabbott commented on static/styles/popovers.css:
```diff
@@ -688,3 +688,29 @@ ul {
     }
 }
+
+.formatting_container {
```
Maybe call this `compose_formatting_button`, to better namespace this logic?

timabbott commented on static/js/popover_menus.js:
```diff
@@ -87,4 +90,48 @@ export function initialize() {
+                // user doesn't click on it by mistake.
+                $popper.find(".tippy-content").css("padding", 0);
+                $popper.on("click", ".formatting_button", (e) => {
+                    const format_type = $(e.currentTarget).find("[data-format]").data("format");
```
Maybe `data-format-type` would be a bit cleaner name for the HTML element containing the `format_type`?

timabbott commented on static/js/popover_menus.js:
```diff
@@ -87,4 +90,48 @@ export function initialize() {
+                    compose_ui.format_text(textarea, format_type);
+                    textarea.trigger("focus");
+                });
+                textarea.on("keyup", () => {
```
Will we accumulate N copies of this handler on the compose textarea after using this feature for a while?  I'm not sure anything removes this when the instance is hidden.

amanagr replied:
Ok, removed it after it's purpose is finished.

timabbott commented on static/js/compose_ui.js:
```diff
@@ -221,17 +221,77 @@ export function handle_keyup(event, textarea) {
+                text.length - range.end >= 2 && // If the characters before and after selected text have bold_syntax.
+                text.slice(range.start - 2, range.start) === bold_syntax &&
+                text.slice(range.end, range.end + 2) === bold_syntax
+            ) {
```
One thing that I notice is that this works if you have `**[selected text]**`, but not `[**selected text**]`, which seems like something a user would do if they were trying to remove the bold a while after adding it in the first place.  GitHub seems to handle that case, so seems like we should too.

The other thing I notice is that this logic could probably be deduplicated with the italics logic; the `2` here is just `syntax.length` and the main thing that differs.

This seems worth doing since we may also decide to add `~~strikethrough~~` and code block wrapping to what's supported by this feature.

amanagr replied:
Fixed it and reduced some duplecacy.

timabbott replied:
Here's a readability refactoring I had in mind, replacing the `1/2` with `.length`, which might also help with reusability:


```diff --git a/static/js/compose_ui.js b/static/js/compose_ui.js
index b1319b2818..460ab4ec98 100644
--- a/static/js/compose_ui.js
+++ b/static/js/compose_ui.js
@@ -231,21 +231,24 @@ export function format_text(textarea, type) {
 
     switch (type) {
         case "bold":
-            // If the text is already bold, we remove the bold_syntax from text.
-            // Check if characters around selected text have enough length to have bold_syntax.
+            // If the selection is already surrounded by bold syntax,
+            // remove it rather than adding another copy.
             if (
-                range.start >= 2 &&
-                text.length - range.end >= 2 && // If the characters before and after selected text have bold_syntax.
-                text.slice(range.start - 2, range.start) === bold_syntax &&
-                text.slice(range.end, range.end + 2) === bold_syntax
+                // First check if there are enough characters before/after selection.
+                range.start >= bold_syntax.length &&
+                text.length - range.end >= bold_syntax.length &&
+                // And then if the characters are bold_syntax.
+                text.slice(range.start - bold_syntax.length, range.start) === bold_syntax &&
+                text.slice(range.end, range.end + bold_syntax.length) === bold_syntax
             ) {
                 // Remove the bold_syntax from text.
                 text =
-                    text.slice(0, range.start - 2) +
+                    text.slice(0, range.start - bold_syntax.length) +
                     text.slice(range.start, range.end) +
-                    text.slice(range.end + 2);
+                    text.slice(range.end + bold_syntax.length);
                 textarea.val(text);
-                field.setSelectionRange(range.start - 2, range.end - 2);
+                field.setSelectionRange(range.start - bold_syntax.length,
+                                        range.end - bold_syntax.length);
                 break;
             }
 

```

amanagr replied:
Done.

timabbott commented:
This looks great!  I merged the first helper commit as 5151e916eae156f3825a9496e2f9cf7453a846d3 and posted a bunch of comments on the implementation.  There's a few weird corner issues I noticed while testing this, mainly involving the fact that Zulip markdown doesn't allow multi-paragraph/bullet bold syntax blocks.  I think it's potentially OK to leave that for follow-up work.

I think the now-first commit and the last commit are probably mergeable to master after the next round of feedback (since they appear to only improve the UX of using the existing keyboard shortcuts); the rest I'd want to test-deploy on chat.zulip.org for feedback before merging.

amanagr commented:
CI is passing. Weird cancelled run.

timabbott commented:
I merged the first commit, which migrates to ` text-field-edit`, as 1347ee98eeafc0723440fec631d901d857427339.

The popover introduction commit has a few outstanding issues:
* I'm not sure we want to have inline PNGs in the HTML; that's not a thing we do as a code pattern, and probably for good reason.
* The popover-hiding logic thing noted above
* We probably want to test it on CZO a bit.

I also note that commit could be moved to the end of the branch, in that it commutes with the undo/trimming logic commits, which are useful independent of the popover, since they fix issues with the keyboard UI.

The undo/trimming commits really feel like they could benefit from node tests in the `compose_ui.js` test suite, at least for the string manipulation logic for determining under what circumstances to trigger an undo.  It feels to me that the logic is quite complex, and without a test suite it'll be hard to maintain and iterate on.

The main bug I noticed with those commits is that the "Ctrl+B should undo adding bold" feature breaks the browser's native undo feature, because it changes the content of the box with `textarea.val`, rather than a `text-field-edit` method (or browser native feature) that is safe for such changes. 

But I also think we may want to think about how to make the logic for the "toggle bold/italics" simpler.  What if we just wrote a helper function that determined the number of copies of a character (`*` for now, but we'd pass `~` for strikethrough in the future) wrapping the current selection, and if it's at least 2, we undo bold, and if it's 1 or at least 3, we undo italic?  I'm not sure I understand the corner cases you're handling, but I would expect that to roughly do what we want?

amanagr commented:
- [x] I'm not sure we want to have inline PNGs in the HTML; that's not a thing we do as a code pattern, and probably for good reason. -> Used icons, I don't know what was the difference, but it worked this time. (I think this was the 4th attempt to use zulip-icons here from me).
- [x] [Popover hiding](https://github.com/zulip/zulip/pull/18922#discussion_r703590921) Since the compose formatting popover is working correctly with just `popovers.hide_all()`, I changed it to as it was.
- [x] I also note that commit could be moved to the end of the branch, in that it commutes with the undo/trimming logic commits, which are useful independent of the popover, since they fix issues with the keyboard UI.
- [x] The main bug I noticed with those commits is that the "Ctrl+B should undo adding bold" feature breaks the browser's native undo feature, because it changes the content of the box with textarea.val, rather than a text-field-edit method (or browser native feature) that is safe for such changes. -> Thanks for noticing.
- [ ] But I also think we may want to think about how to make the logic for the "toggle bold/italics" simpler. What if we just wrote a helper function that determined the number of copies of a character (* for now, but we'd pass ~ for strikethrough in the future) wrapping the current selection, and if it's at least 2, we undo bold, and if it's 1 or at least 3, we undo italic? I'm not sure I understand the corner cases you're handling, but I would expect that to roughly do what we want?
 -> I tried to but  it isn't as simple, there are still a lot of cases when you get down to it.

- [ ] Working on tests.

timabbott commented:
>  Used icons, I don't know what was the difference, but it worked this time. (I think this was the 4th attempt to use zulip-icons here from me).

I wonder if the issue is that webpack may need to be restarted to pick up new files there?

timabbott commented:
I merged the first two commits as the series ending with c750d0e404ac974a8564311487cf32023b5766d8 after a bunch of manual testing; thanks for fixing all those subtle details!  It's pretty fun to toggle 50 times and then hold Ctrl+Z and see it all work.

I'm test-deploying the last commit on chat.zulip.org for feedback.

amanagr commented:
Pushed to fix night mode color issue.

timabbott commented:
Redeployed this on chat.zulip.org.

amanagr commented:
See #20235 for alternative approach. Closing this, we can reopen easily if needed in future.
NOTE TO SELF: Don't delete this branch.
